Presentation: Unable to load related properties using the same relationship path but different instance filters

Repro branch: https://github.com/iTwin/itwinjs-core/tree/presentation/related-properties-instance-filter
Repro:
1. pull the branch
2. `rush install`
3. `rush build -t presentation-full-stack-tests`
4. `cd full-stack-tests/presentation`
5. `npm run test`
6. observe the failure:
   ```diff
   1) Content
        Related Properties
          creates related fields using same relationship but different instance filter:
 
       AssertionError: expected { label: { …(3) }, …(4) } to contain subset { primaryKeys: [ { …(2) } ], …(1) }
       + expected - actual
 
        {
       -  "displayValues": {
       -    "pc_bis_Element_CodeValue/3": "partition1"
       -  }
       -  "label": {
       -    "displayValue": "partition1"
       -    "rawValue": "partition1"
       -    "typeName": "string"
       -  }
       -  "mergedFieldNames": []
          "primaryKeys": [
            {
              "className": "BisCore:PhysicalPartition"
       -      "id": "0x11"
       +      "id": "0x12"
            }
          ]
          "values": {
       -    "pc_bis_Element_CodeValue/3": "partition1"
       +    "pc_bis_Element_CodeValue/3": "partition2"
          }
        }
       
       at Context.<anonymous> (src/frontend/content/RelatedProperties.test.ts:99:100)
   ```

--- a/full-stack-tests/presentation/src/frontend/content/RelatedProperties.test.ts
+++ b/full-stack-tests/presentation/src/frontend/content/RelatedProperties.test.ts
@@ -1,0 +1,108 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import { expect } from "chai";
+import { Guid } from "@itwin/core-bentley";
+import { Content, ContentSpecificationTypes, InstanceKey, KeySet, NestedContentField, NestedContentValue, Ruleset, RuleTypes } from "@itwin/presentation-common";
+import { Presentation } from "@itwin/presentation-frontend";
+import { collect, getFieldByLabel, getFieldsByLabel } from "../../Utils.js";
+import { describeContentTestSuite } from "./Utils.js";
+import { buildTestIModelConnection, insertPhysicalPartition } from "../../IModelSetupUtils.js";
+
+describeContentTestSuite("Related Properties", () => {
+  it("creates related fields using same relationship but different instance filter", async function () {
+    const rootSubjectId = "0x1";
+    const partitionKeys: InstanceKey[] = [];
+    const imodelConnection = await buildTestIModelConnection(this.test!.fullTitle(), async (db) => {
+      const p1 = insertPhysicalPartition({ db, codeValue: "partition1", parentId: rootSubjectId });
+      const p2 = insertPhysicalPartition({ db, codeValue: "partition2", parentId: rootSubjectId });
+      partitionKeys.push(p1, p2);
+    });
+
+    const ruleset: Ruleset = {
+      id: Guid.createValue(),
+      rules: [
+        {
+          ruleType: RuleTypes.Content,
+          specifications: [
+            {
+              specType: ContentSpecificationTypes.ContentInstancesOfSpecificClasses,
+              classes: { schemaName: "BisCore", classNames: ["Subject"] },
+              instanceFilter: `this.ECInstanceId = 1`,
+            },
+          ],
+        },
+        {
+          ruleType: RuleTypes.ContentModifier,
+          class: { schemaName: "BisCore", className: "Subject" },
+          relatedProperties: [
+            {
+              propertiesSource: {
+                relationship: { schemaName: "BisCore", className: "SubjectOwnsPartitionElements" },
+                direction: "Forward",
+              },
+              instanceFilter: `this.ECInstanceId = ${partitionKeys[0].id}`,
+              properties: [
+                {
+                  name: "CodeValue"
+                }
+              ]
+            },
+          ],
+        },
+        {
+          ruleType: RuleTypes.ContentModifier,
+          class: { schemaName: "BisCore", className: "Subject" },
+          relatedProperties: [
+            {
+              propertiesSource: {
+                relationship: { schemaName: "BisCore", className: "SubjectOwnsPartitionElements" },
+                direction: "Forward",
+              },
+              instanceFilter: `this.ECInstanceId = ${partitionKeys[1].id}`,
+              properties: [
+                {
+                  name: "CodeValue"
+                }
+              ]
+            },
+          ],
+        },
+      ],
+    };
+
+    const content = await Presentation.presentation
+      .getContentIterator({
+        imodel: imodelConnection,
+        rulesetOrId: ruleset,
+        descriptor: {},
+        keys: new KeySet(),
+      })
+      .then(async (x) => x && new Content(x.descriptor, await collect(x.items)));
+
+    const partitionFields = getFieldsByLabel(content!.descriptor.fields, "Information Partition");
+    expect(partitionFields.length).to.eq(2);
+
+    const codeValueField1 = getFieldByLabel((partitionFields[0] as NestedContentField).nestedFields, "Code");
+    expect((content?.contentSet[0].values[partitionFields[0].name] as NestedContentValue[])[0]).to.containSubset({
+      primaryKeys: [
+        partitionKeys[0],
+      ],
+      values: {
+        [codeValueField1.name]: "partition1",
+      },
+    });
+
+    const codeValueField2 = getFieldByLabel((partitionFields[1] as NestedContentField).nestedFields, "Code");
+    expect((content?.contentSet[0].values[partitionFields[1].name] as NestedContentValue[])[0]).to.containSubset({
+      primaryKeys: [
+        partitionKeys[1],
+      ],
+      values: {
+        [codeValueField2.name]: "partition2",
+      },
+    });
+  });
+});

